kale/loaddata/few_shot.py

```diff
@@ -0,0 +1,124 @@
+
+    Note:
+        The dataset should be organized as:
+
+        - root
+            - train
+                - class_name 1
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - class_name 2
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - ...
+            - val
+                - class_name 1
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - class_name 2
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - ...
+            - test (optional)
+                - class_name 1
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - class_name 2
+                    - xxx.png
+                    - yyy.png
+                    - ...
+                - ...
+
+    Args:
+        path (string): The root directory of the data.
+        mode (string): The mode of the dataset. It can be 'train', 'val', or 'test'. Default: 'train'.
+        num_support_samples (int): Number of support examples per class in each iteration. It corresponds to K in the N-way-K-shot setting. Default: 5.
+        num_query_samples (int): Number of query examples per class in each iteration. Default: 15.
+        transform (callable, optional): Optional transform to be applied on a sample. Default: None.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        mode: str = "train",
+        num_support_samples: int = 5,
+        num_query_samples: int = 15,
+        transform: Optional[Callable] = None,
+    ):
+        super(NWayKShotDataset, self).__init__()
+
+        self.root = path
+        self.num_support_samples = num_support_samples
+        self.num_query_samples = num_query_samples
+        self.mode = mode
+        self.transform = transform
+        self.images = []  # type: list
+        self.labels = []  # type: list
+        self._load_data()
+
+    def __len__(self):
+        # the length of the dataset is the number of classes
+        return len(self.classes)
+
+    def __getitem__(self, idx):
+        # sampling data for one class
+        image_idx = self._get_idx(idx)
+        images = self._sample_data(image_idx)
+        assert isinstance(images, list)
+        images = torch.stack(images)
+        return images, idx
+
+    def _get_idx(self, idx):
+        # getting the indices of images for one class
+        image_idx = np.random.choice(
+            [i for (i, item) in enumerate(self.labels) if item == idx],
+            self.num_support_samples + self.num_query_samples,
+            replace=False,
+        )
+        return image_idx
+
+    def _sample_data(self, image_idx):
+        # sampling data for one class
```

Comment:
The same as `def __getitem__(self, idx):`?

Reply:
Under N-way-K-shot settings, we will sample data twice. Firstly, we sample N classes. Then, from each class, we will sample K+Q images. I changed the comments here.